File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var replaceValueInTree = function(root) {
    const map = {}
    findSum(root, 0)
    const dummy = new TreeNode()
    dummy.right = root
    convertTree(dummy, -1)
    return root
    
    function convertTree(root, lvl) {
        if (!root) return
        let nextLvlSum = map[lvl+1]
        if (root.left) nextLvlSum -= root.left.val
        if (root.right) nextLvlSum -= root.right.val
        if (root.left) root.left.val = nextLvlSum
        if (root.right) root.right.val = nextLvlSum
        convertTree(root.left, lvl+1)
        convertTree(root.right, lvl+1)
    }
    
    function findSum(root, lvl) {
        if (!root) return
        map[lvl] = map[lvl] || 0
        map[lvl] += root.val
        findSum(root.left, lvl+1)
        findSum(root.right, lvl+1)
    }
};

File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.mp = {}
        
    def find_sum(self, root: Optional[TreeNode], lvl: int) -> None:
        if not root:
            return
        if lvl not in self.mp:
            self.mp[lvl] = 0
        self.mp[lvl] += root.val
        self.find_sum(root.left, lvl+1)
        self.find_sum(root.right, lvl+1)
        
    def convert(self, root: Optional[TreeNode], lvl: int) -> None:
        if not root:
            return
        lvl_sum = self.mp.get(lvl+1, 0)
        if root.left:
            lvl_sum -= root.left.val
        if root.right:
            lvl_sum -= root.right.val
        if root.left:
            root.left.val = lvl_sum
        if root.right:
            root.right.val = lvl_sum
        self.convert(root.left, lvl+1)
        self.convert(root.right, lvl+1)
    
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.find_sum(root, 0)
        dummy = TreeNode(right=root)
        self.convert(dummy, -1)
        return root